// pinus-robot/dist/lib/master/starter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cp = require("child_process");
const fs = require("fs");
const vm = require("vm");
require("cliff");
function run(main, message, clients) {
    if (!clients) {
        clients = ['127.0.0.1'];
        prepare(main, message, clients);
    }
    else {
        prepare(main, message, clients);
    }
}
exports.run = run;
function prepare(main, message, clients) {
    let count = parseInt(message.agent, 10) || 1;
    for (let ipindex in clients) {
        for (let i = 0; i < count; i++) {
            let cmd = `"${process.execPath}" ${main}`;
            let ip = clients[ipindex];
            if (ip === '127.0.0.1') {
                localrun(cmd);
            }
            else {
                sshrun(cmd, ip);
            }
        }
    }
}
exports.prepare = prepare;
function sshrun(cmd, host, callback) {
    let hosts = [host];
    log('Executing ' + cmd.yellow + ' on ' + hosts.join(', ').blue);
    let wait = 0;
    let data = [];
    if (hosts.length > 1) {
        starter.parallelRunning = true;
    }
    hosts.forEach(function (host) {
        wait += 1;
        spawnProcess('ssh', [host, cmd], function (err, out) {
            if (!err) {
                data.push({
                    host: host,
                    out: out
                });
            }
            done(err);
        });
    });
    let error;
    function done(err) {
        error = error || err;
        if (--wait === 0) {
            starter.parallelRunning = false;
            if (error) {
                abort('FAILED TO RUN, return code: ' + error);
            }
            else if (callback) {
                callback(data);
            }
        }
    }
}
exports.sshrun = sshrun;
function localrun(cmd, callback) {
    log('Executing ' + cmd.green + ' locally');
    spawnProcess(cmd, ['', ''], function (err, data) {
        if (err) {
            abort('FAILED TO RUN, return code: ' + err);
        }
        else {
            if (callback)
                callback(data);
        }
    });
}
exports.localrun = localrun;
let starter = new class {
};
function set(key, def) {
    if (typeof def === 'function') {
        starter.__defineGetter__(key, def);
    }
    else {
        starter.__defineGetter__(key, function () {
            return def;
        });
    }
}
exports.set = set;
function load(file) {
    if (!file)
        throw new Error('File not specified');
    log('Executing compile ' + file);
    let code = fs.readFileSync(file).toString();
    vm.runInNewContext(code, starter);
}
exports.load = load;
function abort(msg) {
    log(msg.red);
    // process.exit(1);
}
exports.abort = abort;
let log = console.log;
/**
 *begin notify to run agent
 *
 *
 */
function addBeauty(prefix, buf) {
    let out = prefix + ' ' + buf
        .toString()
        .replace(/\s+$/, '')
        .replace(/\n/g, '\n' + prefix);
    return out.green;
}
function spawnProcess(command, options, callback) {
    let child = null;
    if (!!options[0]) {
        child = cp.spawn(command, options);
    }
    else {
        child = cp.exec(command, options);
    }
    let prefix = command === 'ssh' ? '[' + options[0] + '] ' : '';
    prefix = prefix.grey;
    // child.stderr.on('data', function (chunk) {
    //    log(addBeauty(chunk));
    // });
    let res = [];
    child.stdout.on('data', function (chunk) {
        res.push(chunk.toString());
        log(addBeauty(chunk));
    });
    function addBeauty(buf) {
        return prefix + buf
            .toString()
            .replace(/\s+$/, '')
            .replace(/\n/g, '\n' + prefix);
    }
    child.on('exit', function (code) {
        if (callback) {
            callback(code === 0 ? null : code, res && res.join('\n'));
        }
    });
}
