// pinus-robot/dist/lib/robot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const agent_1 = require("./agent/agent");
const server_1 = require("./master/server");
const http_1 = require("./console/http");
/**
 * export to developer prototype
 *
 * @param {Object} config
 * include deal with master and agent mode
 *
 * param include mode
 *
 */
class Robot {
    constructor(conf) {
        this.master = null;
        this.agent = null;
        this.conf = conf;
    }
    /*
   * run master server
   *
   * @param {String} start up file
   *
   */
    runMaster(mainFile) {
        let conf = {}, master;
        conf.clients = this.conf.clients;
        conf.mainFile = mainFile;
        this.master = new server_1.Server(conf);
        this.master.listen(this.conf.master.port);
        http_1.HTTP_SERVER.start(`http://${this.conf.master.host}:${this.conf.master.port}`, this.conf.master.webport);
    }
    /**
     * run agent client
     *
     * @param {String} script
     *
     */
    runAgent(scriptFile) {
        let conf = {};
        conf.master = this.conf.master;
        conf.scriptFile = scriptFile;
        this.agent = new agent_1.Agent(conf);
        this.agent.start();
    }
    restart() {
        if (this.agent != null) {
            this.agent.reconnect(true);
        }
    }
}
exports.Robot = Robot;
