// pinus-robot/dist/lib/monitor/monitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 *
 * agent monitor data map
 *
 * every agent put start and end time in to route map
 * then report to master
 *
 */
const fs = require("fs");
let util = require('../common/util');
let dataMap = {};
let incrMap = {};
let profData = {};
function getData() {
    return {
        timeData: profData,
        incrData: incrMap
    };
}
exports.getData = getData;
function clear() {
    profData = {};
    incrMap = {};
}
exports.clear = clear;
function incr(name) {
    incrMap[name] = incrMap[name] == null ? 1 : incrMap[name] + 1;
    console.log(incrMap[name] + ' ' + name);
}
exports.incr = incr;
function decr(name) {
    incrMap[name] = incrMap[name] == null ? 0 : incrMap[name] - 1;
}
exports.decr = decr;
function beginTime(route, uid, id) {
    let time = Date.now();
    if (!dataMap[route]) {
        dataMap[route] = buildMapData();
    }
    if (!dataMap[route][uid]) {
        dataMap[route][uid] = buildMapData();
        dataMap[route][uid][id] = time;
    }
    dataMap[route][uid][id] = time;
}
exports.beginTime = beginTime;
function endTime(route, uid, id) {
    if (!dataMap[route]) {
        return;
    }
    if (!dataMap[route][uid]) {
        return;
    }
    if (!dataMap[route][uid][id]) {
        return;
    }
    let beginTime = dataMap[route][uid][id];
    delete dataMap[route][uid][id];
    let span = Date.now() - beginTime;
    // console.log('route span ' + route+ ' ' + uid + ' ' +  span);
    // saveTimes(uid,route+":"+span+'\r\n');
    let srcData = profData[route];
    if (!srcData) {
        srcData = { min: span, max: span, avg: span, num: 1 };
        profData[route] = srcData;
    }
    else {
        if (span < srcData.min) {
            srcData.min = span;
        }
        if (span > srcData.max) {
            srcData.max = span;
        }
        srcData.avg = (srcData.avg * srcData.num + span) / (srcData.num + 1);
        srcData.num = (srcData.num + 1);
    }
}
exports.endTime = endTime;
function buildMapData() {
    let data = {};
    return data;
}
let saveTimes = function (uid, value) {
    fs.appendFile(util.getPath() + '/detail', value, function (err) {
        if (err) {
            console.log(err);
        }
    });
};
