// pinus-robot/dist/lib/common/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// ------------------------------------
// Statistics Manager
// ------------------------------------
//
// This file defines qputs, qprint, and extends the util namespace.
//
// Extends node.js util.js with other common functions.
//
let fs = require('fs');
exports.getPath = function () {
    let path = './';
    if (__filename.indexOf('node_modules') === -1) {
        path = __filename.substring(0, __filename.lastIndexOf('/')) + '/../../log';
    }
    else {
        path = __filename.substring(0, __filename.lastIndexOf('node_modules')) + 'log';
    }
    return path;
};
exports.createPath = function () {
    let path = exports.getPath();
    if (!require('fs').existsSync(path)) {
        fs.mkdirSync(path);
    }
};
exports.deleteLog = function () {
    let path = exports.getPath();
    try {
        fs.unlinkSync(path + '/detail');
        fs.unlinkSync(path + '/.log');
    }
    catch (ex) {
    }
};
// A few common global functions so we can access them with as few keystrokes as possible
//
exports.qputs = function (s) {
    console.log(s);
};
exports.qprint = function (s) {
    console.log(s);
};
// Static utility methods
//
exports.uid = function () {
    exports.lastUid_ = exports.lastUid_ || 0;
    return exports.lastUid_++;
};
exports.defaults = function (obj, defaults) {
    for (let i in defaults) {
        if (obj[i] === undefined) {
            obj[i] = defaults[i];
        }
    }
    return obj;
};
exports.extend = function (obj, extension) {
    for (let i in extension) {
        if (extension.hasOwnProperty(i)) {
            obj[i] = extension[i];
        }
    }
    return obj;
};
exports.forEach = function (obj, f) {
    for (let i in obj) {
        if (obj.hasOwnProperty(i)) {
            f(i, obj[i]);
        }
    }
};
exports.every = function (obj, f) {
    for (let i in obj) {
        if (obj.hasOwnProperty(i)) {
            if (!f(i, obj[i])) {
                return false;
            }
        }
    }
    return true;
};
exports.argarray = function (args) {
    return Array.prototype.slice.call(args);
};
exports.readStream = function (stream, callback) {
    let data = [];
    stream.on('data', function (chunk) {
        data.push(chunk.toString());
    });
    stream.on('end', function () {
        callback(data.join(''));
    });
};
/** Same arguments as http.createClient. Returns an wrapped http.Client object that will reconnect when
connection errors are detected. In the current implementation of http.Client (11/29/10), calls to
request() fail silently after the initial 'error' event. */
exports.createReconnectingClient = function () {
    let http = require('http'), clientArgs = arguments, events = {}, client, wrappedClient = {}, clientMethod = function (method) {
        return function () { return client[method].apply(client, arguments); };
    }, clientGetter = function (member) { return function () { return client[member]; }; }, clientSetter = function (member) { return function (val) { client[member] = val; }; }, reconnect = function () {
        let oldclient = client;
        if (oldclient) {
            oldclient.destroy();
        }
        client = http.createClient.apply(http, clientArgs);
        client._events = exports.extend(events, client._events); // EventEmitter._events stores event handlers
        client.emit('reconnect', oldclient);
    };
    // Create initial http.Client
    reconnect();
    client.on('error', function (err) { reconnect(); });
    // Wrap client so implementation can be swapped out when there are connection errors
    for (let j in client) {
        if (typeof client[j] === 'function') {
            wrappedClient[j] = clientMethod(j);
        }
        else {
            wrappedClient.__defineGetter__(j, clientGetter(j));
            wrappedClient.__defineSetter__(j, clientSetter(j));
        }
    }
    wrappedClient.impl = client;
    return wrappedClient;
};
